### PR TITLE
Implement finding meeting times functionality

### PR DESCRIPTION
**Pull request by shainachen**
1. **Complete query()**
- query() returns time ranges when a meeting can be held. Considers both optional and mandatory attendees with this rule: if one or more time slots exists so that both mandatory and optional attendees can attend, return those time slots. Otherwise, return the time slots that fit just the mandatory attendees.
2. **Write tests to test behavior with optional attendees**

The workflow of this week's assignment was:

1. Write query() to satisfy given tests for mandatory attendees
2. Write tests to test optional attendee behavior
3. Modify query() to satisfy these new tests, so that it works with both mandatory and optional attendees.

See detailed project assignment [here](https://github.com/googleinterns/step/blob/master/walkthroughs/week-5-tdd/project/calendar-walkthrough.md).

### CODE REVIEW

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 1]
```diff
@@ -270,5 +272,159 @@ public void notEnoughRoom() {
...
+    request.addOptionalAttendee(PERSON_C);
+    Collection<TimeRange> expected =
+    Arrays.asList(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TIME_0800AM, false),
+        TimeRange.fromStartEnd(TIME_0830AM, TIME_0900AM, false),
+        TimeRange.fromStartEnd(TIME_0930AM, TimeRange.END_OF_DAY, true));
+
+    //Act
+    Collection<TimeRange> actual = query.query(events, request);
+
+    //Assert
```
Comment (NathanBricault) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:305]: Explicitly calling out the steps is unnecessary; they should generally be self-evident.

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 2]
```diff
@@ -270,5 +272,159 @@ public void notEnoughRoom() {
 
     Assert.assertEquals(expected, actual);
   }
-}
 
+  @Test
+  public void testFindMeetingQuery_query_optionalAttendeeHasConflictingEventAndNotConsidered() {
```
Comment (NathanBricault) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:277]: These test names are a bit hard to understand. Focus on communicating the minimal setup and the expected behavior. It's not a strict style requirement, but a format of <some description of class>_<key detail of setup>_<expected behavior> tends to work well and be a little easier to read.
Maybe something like: testQuery_optionalAttendeeHasConflict_returnsValidTimesForOtherGuests

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 3]
```diff
@@ -270,5 +272,159 @@ public void notEnoughRoom() {
...
+            Arrays.asList(PERSON_B)),
+        new Event("Event 3", TimeRange.fromStartDuration(TimeRange.START_OF_DAY, TimeRange.WHOLE_DAY.duration()),
+            Arrays.asList(PERSON_C)));
+    MeetingRequest request =
+        new MeetingRequest(Arrays.asList(PERSON_A, PERSON_B), DURATION_30_MINUTES);
+    request.addOptionalAttendee(PERSON_C);
+    Collection<TimeRange> expected =
+    Arrays.asList(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TIME_0800AM, false),
+        TimeRange.fromStartEnd(TIME_0830AM, TIME_0900AM, false),
+        TimeRange.fromStartEnd(TIME_0930AM, TimeRange.END_OF_DAY, true));
```
Comment (NathanBricault) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:300]: Optional: whether you consider this part of arrange or part of assert is a bit debatable. I would generally lean towards considering it part of the assert phase, as I like to keep the arrange phase as minimal as possible, including exactly as much code as is the minimum requirement to set the stage.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -14,10 +14,89 @@
...
+  * Considers mandatory and optional attendees. If no time ranges are available to accomodate both
+  * types of attendees, returns time ranges that accomodate just mandatory attendees.
+  * @param events collection of events happening during the day
+  * @param request request for meeting with specified requirements (duration, attendees)
+  * @return collection of time ranges that requested meeting can be held at. 
+  */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<String> mandatoryAttendees = request.getAttendees();
+    Collection<String> optionalAndMandatoryAttendees = Lists.newArrayList(Iterables.unmodifiableIterable(
```
[the remark this answers is not in the document]
Comment (shainachen) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:37]: Changed to a set, thanks for the suggestion

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 5]
```diff
@@ -283,148 +285,136 @@ public void testFindMeetingQuery_query_optionalAttendeeHasConflictingEventAndNot
...
-    //Act
     Collection<TimeRange> actual = query.query(events, request);
 
-    //Assert
     Assert.assertEquals(expected, actual);
   }
 
   @Test
-  public void testFindMeetingQuery_query_everyAttendeeIncludingOptionalOnesAreConsidered() {
+  public void testQuery_optionalAttendeeHasAvailability_returnsValidTimesForMandatoryAndOptionalAttendees() {
```
[the remark this answers is not in the document]
Comment (shainachen) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:310]: Thank you for the feedback on test names!

**Pull request merged**